doc/sphinx/Pacemaker_Development/c.rst

```diff
@@ -15,6 +15,203 @@ Some existing Pacemaker code does not follow these guidelines, for historical
+can package the new version. Therefore, changes to public APIs should be
+backward-compatible (as detailed throughout this chapter), unless we are doing
+a (rare) release where we specifically intend to break compatibility.
+
+External applications known to use Pacemaker's public C API include
+`sbd <https://github.com/ClusterLabs/sbd>`_ and dlm_controld.
+
+
+.. index::
+   pair: C; API documentation
+   single: Doxygen
+
+API Documentation
+_________________
+
+Pacemaker uses `Doxygen <https://www.doxygen.nl/manual/docblocks.html>`_
+to automatically generate its
+`online API documentation <https://clusterlabs.org/pacemaker/doxygen/>`_,
+so all public API (header files, functions, structs, enums, etc.) should be
+documented with Doxygen comment blocks. Other code may be documented in the
+same way if desired, with an ``\internal`` tag in the Doxygen comment.
+
+Simple example of an internal function with a Doxygen comment block:
+
+.. code-block:: c
+
+   /*!
+    * \internal
+    * \brief Return string length plus 1
+    *
+    * Return the number of characters in a given string, plus one.
+    *
+    * \param[in] s  A string (must not be NULL)
+    *
+    * \return The length of \p s plus 1.
+    */
+   static int
+   f(const char *s)
+   {
+      return strlen(s) + 1;
+   }
+
+
+API Header File Naming
+______________________
+
+* Internal API headers should be named ending in ``_internal.h``, in the same
+  location as public headers, with the exception of libpacemaker, which for
+  historical reasons keeps internal headers in ``include/pcmki/pcmki_*.h``).
+
+* If a library needs to share symbols just within the library, header files for
+  these should be named ending in ``_private.h`` and located in the library
+  source directory (not ``include``). Such functions should be declared as
+  ``G_GNUC_INTERNAL``, to aid compiler efficiency (glib defines this
+  symbol appropriately for the compiler).
+
+Header files that are not library API are located in the same locations as
+other source code.
+
+
+.. index::
+   pair: C; naming
+
+API Symbol Naming
+_________________
+
+Exposed API symbols (non-``static`` function names, ``struct`` and ``typedef``
+names in header files, etc.) must begin with the prefix appropriate to the
+library (shown in the table at the beginning of this section). This reduces the
+chance of naming collisions when external software links against the library.
+
+The prefix is usually lowercase but may be all-caps for some defined constants
+and macros.
+
+Public API symbols should follow the library prefix with a single underbar
+(for example, ``pcmk_something``), and internal API symbols with a double
+underbar (for example, ``pcmk__other_thing``).
+
+File-local symbols (such as static functions) and non-library code do not
+require a prefix, though a unique prefix indicating an executable (controld,
+crm_mon, etc.) can be helpful to indicate symbols shared between multiple
+source files for the executable.
+
```

Comment:
Either here or in the previous section, it would be worth mentioning the problem that 2c35da246ae8ecb36d0acd63efad59124f7e23b3 is meant to solve.

Reply:
I actually want to add a whole chapter on coccinelle and unit testing but that will be a separate PR

I'm also thinking a chapter on autotools (configure.ac + Makefile.am) that the little Makefile.am note in this PR would go into

And then chapters on specific chunks of code that merit detail

Reply:
Now that I'm thinking about it, that coccinelle/unit testing chapter would be a great side project for you in between other stuff :)

I'm thinking there's too much stuff for just one C chapter. The current one could be generic guidelines, then we could have additional chapters on special topics. I was thinking of lumping the stuff in devel/ (coccinelle and gdbhelpers) with unit tests in one "stuff to help developers code in C" chapter. Maybe "C Development Helpers" or something like that.